### PR TITLE
Feature/2.0.0

### PR DESCRIPTION
**Pull request by th-anguyen**

### CODE REVIEW

File: cordova-plugin-one/README.md [hunk 1]
```diff
@@ -10,9 +10,12 @@ $ cordova plugin add cordova-plugin-one
 ```
 ## Usage
 ### Initialization
-To initialize the ONE Cordova Plugin, call the following method:
+To initialize the ONE Cordova Plugin, fetch the `One` instance from the window and call the following method:
```
Comment (th-apop) [inline cordova-plugin-one/README.md:13]: What's the reason for the change? Is calling from the window required?

File: cordova-plugin-one/README.md [hunk 2]
```diff
@@ -10,9 +10,12 @@ $ cordova plugin add cordova-plugin-one
 ```
```
Comment (th-apop) [inline cordova-plugin-one/README.md:10]: any plans to add a TOC?

Comment (th-apop) [inline cordova-plugin-one/README.md:10]: Ideally update the whole structure to match order of native methods so it's easier to review if you look at one or the other. Also please add any requirements at the top similar to the native SDKs.

File: cordova-plugin-one/package.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "cordova-plugin-one",
-  "version": "1.4.0",
+  "version": "2.0.0",
```
Comment (th-apop) [inline cordova-plugin-one/package.json:3]: 🚀

File: cordova-plugin-one/src/android/OnePlugin.java [hunk 4]
```diff
@@ -8,419 +8,451 @@
...
+      return initializeOne(args, callbackContext);
+    } else if ("sendInteraction".equals(action)) {
+      cordova.getActivity().runOnUiThread(() -> {
+        try {
+          JSONObject jsonObject = args.optJSONObject(1);
+          HashMap<String, String> propertiesMap = getPropertiesFromJSONObject(jsonObject);
+          sendInteraction(args.getString(0), propertiesMap, callbackContext);
+        } catch (Exception e) {
+          e.printStackTrace();
+          callbackContext.error("Failed to send properties: " + e.getLocalizedMessage());
```
Comment (th-apop) [inline cordova-plugin-one/src/android/OnePlugin.java:57]: Do we need to tag the logs or is it pretty obvious they are coming from our plugin?

File: cordova-plugin-one/src/android/OnePlugin.java [hunk 5]
```diff
@@ -8,419 +8,451 @@
...
-            });
-            return true;
+  protected static final String LOG_TAG = "OnePlugin";
+
+  @Override
+  public boolean execute(final String action, final CordovaArgs args, final CallbackContext callbackContext) throws JSONException {
+    if ("initializeOne".equals(action)) {
+      return initializeOne(args, callbackContext);
+    } else if ("sendInteraction".equals(action)) {
+      cordova.getActivity().runOnUiThread(() -> {
```
Comment (th-apop) [inline cordova-plugin-one/src/android/OnePlugin.java:50]: do we need to `runOnUiThread` for these to work as expected?

File: examples/optimizing-programmatically-using-json-example/README.md [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+   NodeJS            : v15.13.0 (/usr/local/Cellar/node/15.13.0/bin/node)
+   npm               : 7.7.6
+   OS                : macOS Big Sur
+   Xcode             : Xcode 12.4 Build version 12D4e
+
+```
+
+### Setup
+
+1. To create Orchestrations in ONE for this tutorial, see this [guide](https://na5.thunderhead.com/one/help/conversations/how-do-i/mobile/ios-orchestrations/one_integrate_mobile_ios_orch_intro/).
```
Comment (th-apop) [inline examples/optimizing-programmatically-using-json-example/README.md:43]: needs to be updated once we have a bespoke one

File: examples/optimizing-programmatically-using-json-example/README.md [hunk 7]
```diff
@@ -0,0 +1,95 @@
+## Orchestration Example
```
Comment (th-apop) [inline examples/optimizing-programmatically-using-json-example/README.md:1]: I'm not seeing the updates based on our review yesterday?

File: ionic4-angular-example/README.md [hunk 8]
```diff
@@ -12,14 +12,14 @@ $ ionic cordova plugin add cordova-plugin-one
```
Comment (th-apop) [inline ionic4-angular-example/README.md:12]: please ensure the feedback from yesterday's session is also added to the 'old' example app readmes

**Review by th-apop — CHANGES_REQUESTED** [review]

File: cordova-plugin-one/src/android/OnePlugin.java (hunk 4 above)
Comment (th-anguyen) [reply cordova-plugin-one/src/android/OnePlugin.java:57]: Good call, I'll add a tag and use Log.e instead.

File: cordova-plugin-one/README.md (hunk 2 above)
Comment (th-anguyen) [reply cordova-plugin-one/README.md:10]: Yeah I'll add it in there in this go

File: ionic4-angular-example/README.md (hunk 8 above)
Comment (th-anguyen) [reply ionic4-angular-example/README.md]: Updated

File: examples/optimizing-programmatically-using-json-example/README.md (hunk 6 above)
Comment (th-anguyen) [reply examples/optimizing-programmatically-using-json-example/README.md:43]: Have we created a ticket for it yet?

File: cordova-plugin-one/src/android/OnePlugin.java (hunk 5 above)
Comment (th-anguyen) [reply cordova-plugin-one/src/android/OnePlugin.java]: No we do not.  Removed

File: examples/optimizing-programmatically-using-json-example/README.md (hunk 6 above)
Comment (th-apop) [reply examples/optimizing-programmatically-using-json-example/README.md:43]: MOB-1279

File: cordova-plugin-one/README.md [hunk 9]
```diff
@@ -3,16 +3,54 @@
...
+    * [Disable automatic identity transfer](#disable-automatic-identity-transfer)
+    * [Get a URL with one-tid](#get-a-url-with-one-tid)
+    * [Send properties for an incoming URL](#send-properties-for-an-incoming-url)
+    * [Disable automatic outbound link tracking](#disable-automatic-outbound-link-tracking)
+        * [Programmatically trigger an outbound link tracking Interaction call](#programmatically-trigger-an-outbound-link-tracking-interaction-call)
+    * [Get Tid](#get-tid)
+    * [Clear the User Profile](#clear-the-user-profile)
+* [Plugin removal](#plugin-removal)
+
+## Requirements
```
Comment (th-apop) [inline cordova-plugin-one/README.md:28]: should we also call out Ionic / Cordova requirements?

File: cordova-plugin-one/hooks/lib/android-helper.js [hunk 10]
```diff
@@ -7,31 +7,65 @@ module.exports = {
...
-        buildGradle = buildGradle.replace(/\n\s+\/\/ One Cordova Plugin - Start One SDK configuration[\s\S]*\/\/ One Cordova Plugin - End One SDK configuration/, "");
-        utilities.writeBuildGradle(buildGradle);
+      var buildGradle = utilities.readBuildGradle();
+      buildGradle = buildGradle.replace(/\n\s+\/\/ One Cordova Plugin - Start One SDK configuration[\s\S]*\/\/ One Cordova Plugin - End One SDK configuration/, "");
+      utilities.writeBuildGradle(buildGradle);
+    },
+
+    addApplicationToManifest: function (context) {
+      logger.log("One plugin: Adding android:name=\"com.thunderhead.MyApplication\" to AndroidManifest.xml...");
+      var APP_CLASS = 'com.thunderhead.MyApplication';
```
Comment (th-apop) [inline cordova-plugin-one/hooks/lib/android-helper.js:44]: `MyApplication` -> `ThunderheadApplication`

File: cordova-plugin-one/plugin.xml [hunk 11]
```diff
@@ -34,18 +34,38 @@
...
+            </dict>
+        </dict>
+        </config-file>
         <podspec>
             <config>
                 <source url="https://github.com/CocoaPods/Specs.git"/>
             </config>
             <pods use-frameworks="true">
-                <pod name="Thunderhead" spec="~> 5.2.1" />
+                <pod name="Thunderhead" spec="~> 6.3.2" />
```
Comment (th-apop) [inline cordova-plugin-one/plugin.xml:51]: please use the latest version

File: cordova-plugin-one/src/android/MyApplication.java [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+      super.onCreate();
+
+      final OneCodelessInteractionTrackingConfiguration codelessInteractionTrackingConfiguration =
+        new OneCodelessInteractionTrackingConfiguration.Builder()
+          // disables Fragment/Activity Interaction Tracking
+          .disableCodelessInteractionTracking(true)
+          // disables WebView URL Interaction Tracking
+          .disableWebViewInteractionTracking(true)
+          // disables Outbound Link Tracking
+          .disableOutboundLinkTracking(true)
```
Comment (th-apop) [inline cordova-plugin-one/src/android/MyApplication.java:18]: we don't want this to be disabled, please remove

File: cordova-plugin-one/src/android/MyApplication.java [hunk 13]
```diff
@@ -0,0 +1,22 @@
+package com.thunderhead;
+
+import android.app.Application;
+import com.thunderhead.android.api.codeless.OneCodelessInteractionTrackingConfiguration;
+
+public class MyApplication extends Application {
```
Comment (th-apop) [inline cordova-plugin-one/src/android/MyApplication.java:6]: update as mentioned above

File: cordova-plugin-one/src/ios/OnePlugin.m [hunk 14]
```diff
@@ -43,6 +71,7 @@ - (void)sendInteraction:(CDVInvokedUrlCommand*)command
                 CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:error.localizedDescription];
                 [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
             } else {
+                [One processResponse:response];
```
Comment (th-apop) [inline cordova-plugin-one/src/ios/OnePlugin.m:74]: why are we calling this internally?

File: cordova-plugin-one/src/ios/OnePlugin.m [hunk 15]
```diff
@@ -53,13 +82,26 @@ - (void)sendInteraction:(CDVInvokedUrlCommand*)command
...
```
Comment (th-apop) [inline cordova-plugin-one/src/ios/OnePlugin.m:85]: same question as above

File: cordova-plugin-one/src/android/OnePlugin.java [hunk 16]
```diff
@@ -8,419 +8,405 @@
...
+        callbackContext.error("Failed to clear user profile");
+      }
+      return true;
+    } else if ("sendInteractionForOutboundLink".equals(action)) {
+      try {
+        String stringURL = args.optString(0);
+        URL url = new URL(stringURL);
+        sendInteractionForOutboundLink(url);
+      } catch (Exception e) {
+        Log.e(LOG_TAG, "Failed to send interaction for outbound link.", e);
```
Comment (th-apop) [inline cordova-plugin-one/src/android/OnePlugin.java:134]: Some exception use the log tag, some don't. Do we need both? 
1. If so, why? 
2. Why don't add the same setup everywhere?
3. Pull the message in a string and reuse to not duplicate.

File: cordova-plugin-one/README.md (hunk 1 above)
Comment (th-anguyen) [reply cordova-plugin-one/README.md]: By default, yes.  We will have to create an Ionic Native Wrapper in order to fetch ONE natively.

Comment (th-anguyen) [reply cordova-plugin-one/README.md]: Created MOB-1301 to track

File: cordova-plugin-one/src/android/OnePlugin.java (hunk 16 above)
Comment (th-anguyen) [reply cordova-plugin-one/src/android/OnePlugin.java:134]: Missed one other spot that was still using printStackTrace.  Updated all exceptions to Log error with LOG_TAG for consistency.

File: cordova-plugin-one/src/ios/OnePlugin.m (hunk 15 above)
Comment (th-anguyen) [reply cordova-plugin-one/src/ios/OnePlugin.m]: Removed

File: cordova-plugin-one/README.md [hunk 17]
```diff
@@ -24,6 +70,22 @@ To initialize the ONE Cordova Plugin, call the following method:
...
 * See example of usage [here](https://github.com/thunderheadone/one-sdk-cordova/tree/master/ionic5-angular-example/src/app/app.component.ts#L28)
 
+## Additional features
+
+### Opt out
+
+#### Opt an end-user out of tracking
+To opt an end-user out of tracking, when the end-user does not give permission to be tracked in the client app, call the following public method:
+```javascript
+One.optOut(true);
```
Comment (th-skhozin) [inline cordova-plugin-one/README.md:80]: Just a reminder to replace uppercase `One`s by lowercase `one`s.

File: examples/optimizing-programmatically-using-json-example/resources/android/xml/network_security_config.xml [hunk 18]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="utf-8"?>
+<network-security-config>
+    <domain-config cleartextTrafficPermitted="true">
+        <domain includeSubdomains="true">localhost</domain>
+    </domain-config>
+</network-security-config>
```
Comment (th-skhozin) [inline examples/optimizing-programmatically-using-json-example/resources/android/xml/network_security_config.xml:6]: Replace with the following:

`<?xml version="1.0" encoding="utf-8"?>
<network-security-config>
    <base-config>
        <trust-anchors>
            <certificates src="user" />
            <certificates src="system" />
        </trust-anchors>
    </base-config>
</network-security-config>`

**Review by th-skhozin — APPROVED** [review]
Approving, assuming a note will be added about running from the terminal for the first time

File: cordova-plugin-one/README.md [hunk 19]
```diff
@@ -3,42 +3,123 @@
...
+    * [Disable automatic identity transfer](#disable-automatic-identity-transfer)
+    * [Get a URL with one-tid](#get-a-url-with-one-tid)
+    * [Send properties for an incoming URL](#send-properties-for-an-incoming-url)
+    * [Disable automatic outbound link tracking](#disable-automatic-outbound-link-tracking)
+        * [Programmatically trigger an outbound link tracking Interaction call](#programmatically-trigger-an-outbound-link-tracking-interaction-call)
+    * [Get Tid](#get-tid)
+    * [Clear the User Profile](#clear-the-user-profile)
+* [Plugin removal](#plugin-removal)
+
+## Requirements
```
Comment (th-apop) [inline cordova-plugin-one/README.md:31]: 💯

File: cordova-plugin-one/src/android/OneSdkVersion.java [hunk 20]
```diff
@@ -1,18 +1,12 @@
...
-            SdkExtraParams sdkExtraParams = new SdkExtraParams();
-            sdkExtraParams.putStringExtraParameter(SdkExtraParams.PluginsApiParametersKeys.PARAMETER_CORDOVA_VERSION, ONE_CORDOVA_PLUGIN_VERSION);
-            one.passExtraParameters(sdkExtraParams);
-        } else {
-            ThunderheadLogger.errorLog("Executing updateModuleSdkVersion(One) while one is null");
-        }
-    }
-}
\ No newline at end of file
+  private static final String ONE_CORDOVA_PLUGIN_VERSION = "2.0.0";
```
Comment (th-apop) [inline cordova-plugin-one/src/android/OneSdkVersion.java:4]: we should add this to the release process

**Review by th-apop — APPROVED** [review]
